python/lsst/ts/standardscripts/maintel/closed_loop.py

```diff
@@ -0,0 +1,519 @@
+
+        intra_image = await self.latiss.take_cwfs(
+            exptime=self.exposure_time,
+            n=1,
+            group_id=self.group_id,
+            filter=self.filter,
+            grating=self.grating,
+            reason="INTRA" + ("" if self.reason is None else f"_{self.reason}"),
+            program=self.program,
+        )
+
+        self.log.debug("Moving to extra-focal position")
+
+        # Hexapod offsets are relative, so need to move 2x the offset
+        # to get from the intra- to the extra-focal position.
+        # then add the offset to compensate for un-equal magnification
+        z_offset = -(self.dz * 2.0 + self.extra_focal_offset)
+        await self.mtcs.move_camera_hexapod(z=z_offset)
+
+        self.log.debug("Taking extra-focal image")
+
+        extra_image = await self.latiss.take_cwfs(
+            exptime=self.exposure_time,
+            n=1,
+            group_id=self.group_id,
+            filter=self.filter,
+            grating=self.grating,
+            reason="EXTRA" + ("" if self.reason is None else f"_{self.reason}"),
+            program=self.program,
+        )
+
+        return intra_image, extra_image
+
+    async def arun(self, checkpoint: bool = False) -> None:
+        """Perform CWFS measurements and hexapod adjustments until the
+        thresholds are reached.
+
+        Parameters
+        ----------
+        checkpoint : `bool`, optional
+            Should issue checkpoints (default=False)?
+
+        Raises
+        ------
+        RuntimeError:
+            If coordinates are malformed.
+        """
+
+        for i in range(self.max_iter):
+            self.log.debug(f"CWFS iteration {i + 1} starting...")
+
+            if checkpoint:
+                await self.checkpoint(
+                    f"[{i + 1}/{self.max_iter}]: CWFS loop starting..."
+                )
+
+                await self.checkpoint(f"[{i + 1}/{self.max_iter}]: Taking image...")
+
+            # Setting visit_id's to none so run_cwfs will take a new dataset.
+            if self.mode == ClosedLoopMode.CWFS:
+                image = await self.lsstcam.take_acq(
+                    self.exposure_time,
+                    group_id=self.group_id,
+                    reason="INFOCUS"
+                    + ("" if self.reason is None else f"_{self.reason}"),
+                    program=self.program,
+                    filter=self.filter,
+                    grating=self.grating,
+                )
+                visit_id = int(image[0])
+
+                self.mtaos.rem.mtaos.cmd_runWEP.set_start(visitId=visit_id)
+            else:
+                intra_image, extra_image = await self.take_intra_extra_focal_images()
+
+                intra_visit_id = int(intra_image[0])
+                extra_visit_id = int(extra_image[0])
+
+                self.mtaos.rem.mtaos.cmd_runWEP.set_start(
+                    visitId=intra_visit_id, extraId=extra_visit_id
+                )
+
```

Comment:
Please, extract the following code into methods that explain what each step does.